packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-bucket-origin.ts

```diff
@@ -0,0 +1,197 @@
+import { Construct } from 'constructs';
+import * as cloudfront from '../../aws-cloudfront';
+import { AccessLevel } from '../../aws-cloudfront';
+import * as iam from '../../aws-iam';
+import { IBucket } from '../../aws-s3';
+import { Annotations, Aws, Names, Stack } from '../../core';
+
+const BUCKET_ACTIONS = {
+  READ: ['s3:GetObject'],
+  WRITE: ['s3:PutObject'],
+  DELETE: ['s3:DeleteObject'],
+};
+
+/**
+ * Properties for configuring a origin using a standard S3 bucket
+ */
+export interface S3BucketOriginBaseProps extends cloudfront.OriginProps {}
+
+/**
+ * Properties for configuring a S3 origin with OAC
+ */
+export interface S3BucketOriginWithOACProps extends S3BucketOriginBaseProps {
+  /**
+  * An optional Origin Access Control
+  * @default - an Origin Access Control will be created.
+  */
+  readonly originAccessControl?: cloudfront.IOriginAccessControl;
+
+  /**
+   * The level of permissions granted in the bucket policy and key policy (if applicable)
+   * to the CloudFront distribution.
+   * @default AccessLevel.READ
+   */
+  readonly originAccessLevels?: AccessLevel[];
+}
+
+/**
+ * Properties for configuring a S3 origin with OAI
+ */
+export interface S3BucketOriginWithOAIProps extends S3BucketOriginBaseProps {
+  /**
+  * An optional Origin Access Identity
+  * @default - an Origin Access Identity will be created.
+  */
+  readonly originAccessIdentity?: cloudfront.IOriginAccessIdentity;
+}
+
+/**
+ * A S3 Bucket Origin
+ */
+export abstract class S3BucketOrigin extends cloudfront.OriginBase {
+  /**
+   * Create a S3 Origin with Origin Access Control (OAC) configured
+   */
+  public static withOriginAccessControl(bucket: IBucket, props?: S3BucketOriginWithOACProps): cloudfront.IOrigin {
+    return new class extends S3BucketOrigin {
+      private originAccessControl?: cloudfront.IOriginAccessControl;
+
+      constructor() {
+        super(bucket, { ...props });
+        this.originAccessControl = props?.originAccessControl;
+      }
+
+      public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+        if (!this.originAccessControl) {
+          this.originAccessControl = new cloudfront.S3OriginAccessControl(scope, 'S3OriginAccessControl');
+        }
+
+        const distributionId = options.distributionId;
+        const actions = this.getActions(props?.originAccessLevels ?? [cloudfront.AccessLevel.READ]);
+        const result = this.grantDistributionAccessToBucket(distributionId, actions);
+
+        // Failed to update bucket policy, assume using imported bucket
+        if (!result.statementAdded) {
+          Annotations.of(scope).addWarningV2('@aws-cdk/aws-cloudfront-origins:updateBucketPolicy',
+            'Cannot update bucket policy of an imported bucket. Set overrideImportedBucketPolicy to true or update the policy manually instead.');
```

Comment:
I searched and cannot find info on `overrideImportedBucketPolicy`. Was the plan to have it as a feature flag?

Imo, we should just throw the warning as feature flag adds to the code complexity.

Reply:
Ah yes it was part of the previous design, I'll remove since we are not updating policies of imported buckets.